Create develop branch after initializing the git repository

oltionchampari opened the pull request:
Closes    phovea/generator-phovea#209

### Summary of changes
After initializing the git repository 


-   stage all changes 

`this.spawnCommandSync('git', ['add', '--all'])`

-    commit all staged changes 

`this.spawnCommandSync('git', ['commit', '-am', '"Initial commit"'])`
  

-  create and switch to develop branch 

 `this.spawnCommandSync('git', ['checkout', '-b', 'develop'])`

oltionchampari committed:
added additional git commands
to init-app generator

phovea/generator-phovea#209

oltionchampari committed:
set cwd excplicitly

phovea/generator-phovea#209

thinkh committed:
Merge branch 'develop' into ochampari/209_create-develop-branch-after-init

thinkh commented on generators/_node/index.js:
```diff
@@ -143,9 +143,19 @@ class PackageJSONGenerator extends Base {
   }
 
   end() {
+    this.log(this.destinationPath())
```
Can that log be removed? Otherwise extend it with a propper title.

thinkh replied:
@oltionchampari Still open...

thinkh commented on generators/_node/index.js:
```diff
@@ -143,9 +143,19 @@ class PackageJSONGenerator extends Base {
     });
+    this.spawnCommandSync('git', ['add', '--all'], {
+      cwd: this.destinationPath()
+    })
```
Missing semicolon.

thinkh commented on generators/_node/index.js:
```diff
@@ -143,9 +143,19 @@ class PackageJSONGenerator extends Base {
+    })
+    this.spawnCommandSync('git', ['commit', '-am', '"Initial commit"'], {
+      cwd: this.destinationPath()
+    })
```
Missing semicolon.

thinkh commented on generators/_node/index.js:
```diff
@@ -143,9 +143,19 @@ class PackageJSONGenerator extends Base {
+    })
+    this.spawnCommandSync('git', ['checkout', '-b', 'develop'], {
+      cwd: this.destinationPath()
+    })
```
Missing semicolon.

thinkh reviewed (changes requested):
In general it works. However, I get some warnings on my Linux about line endings (which we could ignore as it depends on my local git settings) and the mode of the created files is changed **after** the initial commit. The problem is that I have 20 changed files that I need to stage and commit manually. 

```
➜  _temp yo phovea
? Plugin Type slib
? Included Modules (Press <space> to select, <a> to toggle all, <i> to invert selection)phovea_server
? Included Libraries (Press <space> to select, <a> to toggle all, <i> to invert selection)
? Plugin Name temp
? Description
? Project homepage url https://phovea.caleydo.org
? Author's Name Tester
? Author's Email test@test.at
? Author's Homepage
? GitHub username or organization phovea
   create package.json
   create deploy/docker-compose.partial.yml
   create temp/config.json   
   create requirements.txt   
   create docker_packages.txt
   create .gitattributes
   create buildPython.js
   create requirements_dev.txt
   create setup.cfg
   create setup.py
   create tox.ini 
   create .circleci/config.yml
   create docs/conf.py
   create docs/_static/touch.txt
   create docs/_templates/touch.txt
   create docs/index.rst
   create temp/__init__.py
   create .gitignore      
   create docs/.gitignore
   create README.md
   create .editorconfig
   create ISSUE_TEMPLATE.md
   create LICENSE

Your Node.js version 12.13.0 (npm: 6.12.0) matches the required Node.js version 12 (npm: 6). The phovea generator should work as expected.


Initialized empty Git repository in /d/w/datavisyn/_temp/.git/
warning: CRLF will be replaced by LF in .editorconfig.
The file will have its original line endings in your working directory.
warning: CRLF will be replaced by LF in .gitattributes.
The file will have its original line endings in your working directory.
warning: CRLF will be replaced by LF in .gitignore.
The file will have its original line endings in your working directory.
warning: CRLF will be replaced by LF in .vscode/settings.json.
The file will have its original line endings in your working directory.
warning: CRLF will be replaced by LF in ISSUE_TEMPLATE.md.
The file will have its original line endings in your working directory.
warning: CRLF will be replaced by LF in LICENSE.
The file will have its original line endings in your working directory.
warning: CRLF will be replaced by LF in buildPython.js.
The file will have its original line endings in your working directory.
warning: CRLF will be replaced by LF in deploy/docker-compose.partial.yml.
The file will have its original line endings in your working directory.
warning: CRLF will be replaced by LF in docs/.gitignore.
The file will have its original line endings in your working directory.
warning: CRLF will be replaced by LF in docs/conf.py.
The file will have its original line endings in your working directory.
warning: CRLF will be replaced by LF in docs/index.rst.
The file will have its original line endings in your working directory.
warning: CRLF will be replaced by LF in setup.cfg.
The file will have its original line endings in your working directory.
warning: CRLF will be replaced by LF in temp/__init__.py.
The file will have its original line endings in your working directory.
[master (root-commit) 4744d0b] "Initial commit"
 25 files changed, 782 insertions(+)
 create mode 100755 .circleci/config.yml
 create mode 100755 .editorconfig
 create mode 100755 .gitattributes
 create mode 100755 .gitignore
 create mode 100755 .vscode/settings.json
 create mode 100644 .yo-rc.json
 create mode 100755 ISSUE_TEMPLATE.md
 create mode 100755 LICENSE
 create mode 100755 README.md
 create mode 100755 buildPython.js
 create mode 100755 deploy/docker-compose.partial.yml
 create mode 100644 docker_packages.txt
 create mode 100755 docs/.gitignore
 create mode 100755 docs/_static/touch.txt
 create mode 100755 docs/_templates/touch.txt
 create mode 100755 docs/conf.py
 create mode 100755 docs/index.rst
 create mode 100644 package.json
 create mode 100644 requirements.txt
 create mode 100755 requirements_dev.txt
 create mode 100755 setup.cfg
 create mode 100755 setup.py
 create mode 100755 temp/__init__.py
 create mode 100644 temp/config.json
 create mode 100755 tox.ini
Switched to a new branch 'develop'
```

![grafik](https://user-images.githubusercontent.com/5851088/75569751-05d8c100-5a56-11ea-9d4a-6273fc37ad50.png)

oltionchampari commented:
> In general it works. However, I get some warnings on my Linux about line endings (which we could ignore as it depends on my local git settings) and the mode of the created files is changed **after** the initial commit. The problem is that I have 20 changed files that I need to stage and commit manually.
> 
> ```
> ➜  _temp yo phovea
> ? Plugin Type slib
> ? Included Modules (Press <space> to select, <a> to toggle all, <i> to invert selection)phovea_server
> ? Included Libraries (Press <space> to select, <a> to toggle all, <i> to invert selection)
> ? Plugin Name temp
> ? Description
> ? Project homepage url https://phovea.caleydo.org
> ? Author's Name Tester
> ? Author's Email test@test.at
> ? Author's Homepage
> ? GitHub username or organization phovea
>    create package.json
>    create deploy/docker-compose.partial.yml
>    create temp/config.json   
>    create requirements.txt   
>    create docker_packages.txt
>    create .gitattributes
>    create buildPython.js
>    create requirements_dev.txt
>    create setup.cfg
>    create setup.py
>    create tox.ini 
>    create .circleci/config.yml
>    create docs/conf.py
>    create docs/_static/touch.txt
>    create docs/_templates/touch.txt
>    create docs/index.rst
>    create temp/__init__.py
>    create .gitignore      
>    create docs/.gitignore
>    create README.md
>    create .editorconfig
>    create ISSUE_TEMPLATE.md
>    create LICENSE
> 
> Your Node.js version 12.13.0 (npm: 6.12.0) matches the required Node.js version 12 (npm: 6). The phovea generator should work as expected.
> 
> 
> Initialized empty Git repository in /d/w/datavisyn/_temp/.git/
> warning: CRLF will be replaced by LF in .editorconfig.
> The file will have its original line endings in your working directory.
> warning: CRLF will be replaced by LF in .gitattributes.
> The file will have its original line endings in your working directory.
> warning: CRLF will be replaced by LF in .gitignore.
> The file will have its original line endings in your working directory.
> warning: CRLF will be replaced by LF in .vscode/settings.json.
> The file will have its original line endings in your working directory.
> warning: CRLF will be replaced by LF in ISSUE_TEMPLATE.md.
> The file will have its original line endings in your working directory.
> warning: CRLF will be replaced by LF in LICENSE.
> The file will have its original line endings in your working directory.
> warning: CRLF will be replaced by LF in buildPython.js.
> The file will have its original line endings in your working directory.
> warning: CRLF will be replaced by LF in deploy/docker-compose.partial.yml.
> The file will have its original line endings in your working directory.
> warning: CRLF will be replaced by LF in docs/.gitignore.
> The file will have its original line endings in your working directory.
> warning: CRLF will be replaced by LF in docs/conf.py.
> The file will have its original line endings in your working directory.
> warning: CRLF will be replaced by LF in docs/index.rst.
> The file will have its original line endings in your working directory.
> warning: CRLF will be replaced by LF in setup.cfg.
> The file will have its original line endings in your working directory.
> warning: CRLF will be replaced by LF in temp/__init__.py.
> The file will have its original line endings in your working directory.
> [master (root-commit) 4744d0b] "Initial commit"
>  25 files changed, 782 insertions(+)
>  create mode 100755 .circleci/config.yml
>  create mode 100755 .editorconfig
>  create mode 100755 .gitattributes
>  create mode 100755 .gitignore
>  create mode 100755 .vscode/settings.json
>  create mode 100644 .yo-rc.json
>  create mode 100755 ISSUE_TEMPLATE.md
>  create mode 100755 LICENSE
>  create mode 100755 README.md
>  create mode 100755 buildPython.js
>  create mode 100755 deploy/docker-compose.partial.yml
>  create mode 100644 docker_packages.txt
>  create mode 100755 docs/.gitignore
>  create mode 100755 docs/_static/touch.txt
>  create mode 100755 docs/_templates/touch.txt
>  create mode 100755 docs/conf.py
>  create mode 100755 docs/index.rst
>  create mode 100644 package.json
>  create mode 100644 requirements.txt
>  create mode 100755 requirements_dev.txt
>  create mode 100755 setup.cfg
>  create mode 100755 setup.py
>  create mode 100755 temp/__init__.py
>  create mode 100644 temp/config.json
>  create mode 100755 tox.ini
> Switched to a new branch 'develop'
> ```
> 
> ![grafik](https://user-images.githubusercontent.com/5851088/75569751-05d8c100-5a56-11ea-9d4a-6273fc37ad50.png)

This should be no problem you can turn it off in your git config file https://stackoverflow.com/questions/5834014/lf-will-be-replaced-by-crlf-in-git-what-is-that-and-is-it-important and is actually wanted since in our `.gitattributes ` we always want to normalize text line endings to` linux(lf)`.
The reason it happens is because when you create a new plugin the template files are copied 
with `crlf (windows)` line endings and then when you stage the `.gitattributes` file of the plugin you just created it takes effect and tries to normalize (to `lf line endings`) all text files, thus the warning.

thinkh committed:
Merge branch 'develop' into ochampari/209_create-develop-branch-after-init

# Conflicts:
#	generators/_node/index.js

thinkh reviewed (changes requested):
Here are my test result from the last round.

## Chmod on Windows

It seems to be a "problem" with WSL2 and the different chmod in Linux. 

Windows Powershell
```
PS Microsoft.PowerShell.Core\FileSystem::\\wsl$\Ubuntu\home\user\w\test\myplugin>git status
On branch develop
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   .circleci/config.yml
        modified:   .gitattributes
        modified:   ISSUE_TEMPLATE.md
        modified:   LICENSE
        modified:   README.md
        modified:   buildInfo.js
        modified:   index.js
        modified:   src/index.ts
        modified:   tests.webpack.js
        modified:   tsconfig_dev.json
        modified:   tslint.json
        modified:   typedoc.json

no changes added to commit (use "git add" and/or "git commit -a")
```

I can also still see this changes in Git Extensions.

Same directory in Linux (WSL2) looks fine:

```sh
$ git status
On branch develop
nothing to commit, working tree clean
```

So I would just keep it for now.

## Create plugin from workspace

With #370 we can create plugins from the workspace directory. It seems that the newly added git commands do not work with this feature.

### Steps to reproduce

1. `cd workspace`
2. `yo phovea:init-lib`
3. Follow wizard

### Observed behavior

```
...
   create bar/ISSUE_TEMPLATE.md
   create bar/LICENSE


Next steps: 
 npm install
 npm build

Your Node.js version 12.13.1 (npm: 6.12.1) matches the required Node.js version 12 (npm: 6). The phovea generator should work as expected.


/home/holger/w/phovea/test
Initialized empty Git repository in /home/holger/w/phovea/test/bar/.git/
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
    force .yo-rc.json
    force bar/.yo-rc.json
```

### Expected behavior

The git commands should consider the workspace directory.

oltionchampari committed:
Remove console.log

phovea/generator-phovea#381

oltionchampari committed:
Execute git commands  in correct directory

phovea/generator-phovea#209

oltionchampari commented:
Added the `this.cwd` parameter to the git commands which does the job.

```
Initialized empty Git repository in /home/ochampari/workspace/ordino/ordino/asd/.git/
[master (root-commit) 0f65b31] "Initial commit"
 26 files changed, 1518 insertions(+)
 create mode 100644 .circleci/config.yml
 create mode 100644 .editorconfig
 create mode 100644 .gitattributes
 create mode 100644 .github/ISSUE_TEMPLATE/bug-report.md
 create mode 100644 .github/ISSUE_TEMPLATE/feature-request.md
 create mode 100644 .github/PULL_REQUEST_TEMPLATE.md
 create mode 100644 .github/PULL_REQUEST_TEMPLATE/release.md
 create mode 100644 .gitignore
 create mode 100644 ISSUE_TEMPLATE.md
 create mode 100644 LICENSE
 create mode 100644 README.md
 create mode 100644 buildInfo.js
 create mode 100644 index.js
 create mode 100644 jest.config.js
 create mode 100644 package.json
 create mode 100644 phovea_registry.js
 create mode 100644 src/index.ts
 create mode 100644 src/phovea.ts
 create mode 100644 tests.webpack.js
 create mode 100644 tests/index.test.ts
 create mode 100644 tsconfig.json
 create mode 100644 tsconfig_dev.json
 create mode 100644 tsd.d.ts
 create mode 100644 tslint.json
 create mode 100644 typedoc.json
 create mode 100644 webpack.config.js
Switched to a new branch 'develop'
    force .yo-rc.json
    force asd/.yo-rc.json


```

anita-steiner commented on generators/_node/index.js:
```diff
@@ -173,6 +173,16 @@ class PackageJSONGenerator extends Base {
+    this.spawnCommandSync('git', ['add', '--all'], {
+      cwd: this.destinationPath(this.cwd)
+    });
+    this.spawnCommandSync('git', ['commit', '-am', '"Initial commit"'], {
```
I think the correct statement would be 'git commit --allow-empty -m 'Initial commit'

anita-steiner commented on generators/_node/index.js:
```diff
@@ -173,6 +173,16 @@ class PackageJSONGenerator extends Base {
+    this.spawnCommandSync('git', ['checkout', '-b', 'develop'], {
+      cwd: this.destinationPath(this.cwd)
+    });
+
```
please commit also the empty develop branch
